presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -340,28 +340,26 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
 
         TableMetadata tableMetadata = metadata.getTableMetadata(session, insert.getTarget());
 
-        List<ColumnMetadata> visibleTableColumns = tableMetadata.getColumns().stream()
-                .filter(column -> !column.isHidden())
-                .collect(toImmutableList());
-        List<String> visibleTableColumnNames = visibleTableColumns.stream()
-                .map(ColumnMetadata::getName)
-                .collect(toImmutableList());
-
         RelationPlan plan = createRelationPlan(analysis, insertStatement.getQuery());
 
         Map<String, ColumnHandle> columns = metadata.getColumnHandles(session, insert.getTarget());
         Assignments.Builder assignments = Assignments.builder();
+        boolean supportsMissingColumnsOnInsert = metadata.supportsMissingColumnsOnInsert(session, insert.getTarget());
+        ImmutableList.Builder<ColumnHandle> columnsToBeInserted = ImmutableList.builder();
+        ImmutableList.Builder<ColumnMetadata> visibleColumnsBuilder = ImmutableList.builder();
+
         for (ColumnMetadata column : tableMetadata.getColumns()) {
             if (column.isHidden()) {
                 continue;
             }
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
             int index = insert.getColumns().indexOf(columns.get(column.getName()));
-            if (index < 0) {
+            if (index < 0 && !supportsMissingColumnsOnInsert) {
```

Comment:
when `!supportsMissingColumnsOnInsert` shouldn't you `columnsToBeInserted.add(columns.get(column.getName()))`?

That way you pass all table columns to `io.prestosql.metadata.Metadata#beginInsert`. In general it does matter, but in case of table `x` with columns `col1` and `col2` when  `!supportsMissingColumnsOnInsert` `beginInsert` would be equivalent for statements like `INSERT INTO x (col1) SELECT 1` and `INSERT INTO x (col1, col2) SELECT 1, null`.